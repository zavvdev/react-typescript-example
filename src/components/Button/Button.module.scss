@import "src/styles/theme";

.root {
  background-color: $cPrimary;
  border: none;
  color: $cWhite;
  padding: 10px 20px;
  font-size: 20px;
  font-weight: 500;
  border-radius: 5px;
  cursor: pointer;
  user-select: none;
  display: flex;
  justify-content: center;
  align-items: center;
  text-decoration: none;
  width: max-content;

  &:hover {
    background-color: $cPrimaryDark;
  }
}

.icon {
  margin-right: 10px;
  fill: $cWhite;
  width: 25px;
  display: flex;
  align-items: center;
  margin-left: -3px;
}

.disabled {
  cursor: default;
  pointer-events: none;
  background-color: $cPrimaryLight;
}