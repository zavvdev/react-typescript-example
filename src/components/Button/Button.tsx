import cn from 'classnames';
import classes from 'src/components/Button/Button.module.scss';
import { ButtonProps } from 'src/components/Button/Button.types';

const Button = (props: ButtonProps) => {
  const {
    children,
    className,
    onClick,
    disabled,
    href = undefined,
    target = undefined,
    icon = undefined,
  } = props;

  const rootClasses = cn(classes.root, {
    [classes.disabled]: disabled,
  }, className);

  const renderIcon = () => {
    if (icon) {
      return <div className={classes.icon}>{icon}</div>;
    }
    return null;
  };

  if (href) {
    return (
      <a
        href={href}
        target={target}
        className={rootClasses}
      >
        {renderIcon()}
        {children}
      </a>
    );
  }

  return (
    <button
      type="button"
      disabled={disabled}
      className={rootClasses}
      onClick={onClick}
    >
      {renderIcon()}
      {children}
    </button>
  );
};

export default Button;
