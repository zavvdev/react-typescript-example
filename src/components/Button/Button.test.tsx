import { fireEvent, screen } from '@testing-library/react';
import render from 'src/tests/utils/render';
import Button from 'src/components/Button/Button';

const textMock = 'text_mock';
const hrefMock = 'https://mock.com';

describe('<Button />', () => {
  test('should render Button component (snapshot)', () => {
    const { asFragment } = render(<Button>{textMock}</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should have a text', () => {
    render(<Button>{textMock}</Button>);
    expect(screen.getByText(textMock)).toBeInTheDocument();
  });

  test('should trigger onClick event', () => {
    const onClickFnMock = jest.fn();
    render(
      <Button onClick={onClickFnMock}>
        {textMock}
      </Button>,
    );
    const btn = screen.getByText(textMock);
    fireEvent.click(btn);
    expect(onClickFnMock).toBeCalledTimes(1);
  });

  test('should have icon', () => {
    const iconMock = 'icon_mock';
    render(<Button icon={iconMock}>{textMock}</Button>);
    expect(screen.getByText(iconMock)).toBeInTheDocument();
  });

  test('should render link', () => {
    render(<Button href={hrefMock}>{textMock}</Button>);
    const btn = screen.getByText(textMock);
    expect(btn.getAttribute('href')).toBe(hrefMock);
  });

  test('should render link with target=_blank', () => {
    render(
      <Button
        href={hrefMock}
        target="_blank"
      >
        {textMock}
      </Button>,
    );
    const btn = screen.getByText(textMock);
    expect(btn.getAttribute('target')).toBe('_blank');
  });

  test('should have disabled class', () => {
    render(<Button disabled>{textMock}</Button>);
    const btn = screen.getByText(textMock);
    expect(btn.classList.contains('disabled')).toBeTruthy();
  });
});
