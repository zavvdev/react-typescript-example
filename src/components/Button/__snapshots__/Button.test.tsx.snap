// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render Button component (snapshot) 1`] = `
<DocumentFragment>
  <button
    class="root"
    type="button"
  >
    text_mock
  </button>
</DocumentFragment>
`;
